storage/remote/sigv4.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	signer "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/prometheus/prometheus/config"
+)
+
+type sigV4RoundTripper struct {
+	region string
+	next   http.RoundTripper
+	pool   sync.Pool
+
+	signer *signer.Signer
+}
+
+// NewSigV4RoundTripper returns a new http.RoundTripper that will sign requests
+// using Amazon's Signature Verification V4 signing procedure. The request will
+// then be handed off to the next RoundTripper provided by next. If next is nil,
+// http.DefaultTransport will be used.
+//
+// Credentials for signing are retrieved using the the default AWS credential
+// chain. If credentials cannot be found, an error will be returned.
+func NewSigV4RoundTripper(cfg *config.SigV4Config, next http.RoundTripper) (http.RoundTripper, error) {
```

Comment:
Should this round tripper have header allow list support as mentioned in https://github.com/grafana/grafana/issues/29566? I think it's better if it does.

Reply:
I am not sure about this one, I guess preferably you would talk TLS with amazon directly, therefore nothing should touch your headers :)

Reply:
Oh yes, for sure when you talk to Amazon TLS is required, so your header shouldn't be touched by man in the middle. However, the specific example I have in mind (and encountered) is the Amazon server change the value of the header before sending it off to auth service. For example, the client might have set `x-envoy-timeout-ms` to `10` and sign the header. However, when the request reaches the reverse proxy, the reverse (maybe envoy too) decided `x-envoy-timeout-ms` user set is too big and set it to `5` then send the request off to the actual service. 

When the actual service get the request, authentication will fail because the request was originally signed with value of `10`. So there are some headers that are supposed to be "dynamic" and shouldn't be signed. One common example would be `x-forwarded-for`.

Reply:
Are the minimum headers to sign documented somewhere?

Reply:
The min to sign is the `Host` header. See https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html search for the string `For HTTP/1.1 requests, you must include the host header at a minimum`

Reply:
AWS's own sigv4 proxy goes for a [denylist instead](https://github.com/awslabs/aws-sigv4-proxy/blob/9b9e2fb017638e67973b1b514ae646bdf57326ab/handler/proxy_client.go#L125-L129), but by default it's empty. Not clear which is preferable here, though a denylist is probably more flexible. 

(OTOH, why would Prometheus ever set `X-Forwarded-For`?)

Reply:
I can't think of a reason why Prometheus server would ever set `X-Forwarded-For` :-) I was just using it as an example for "dynamic" header.

I would think a allowlist would be better than an denylist because using denylist is more reactive to problems like "oh, sigv4 is not working... oh, it is because this header value was changed by reverse proxy, let me add it to deny list".  Using allowlist, things are simpler, it either works or doesn't work.  

I think it would be simpler if Prometheus just sign 2 headers: `Host` and `X-Prometheus-Remote-Write-Version`

Reply:
In this case, my suggestion is to blacklist `Uber-Trace-Id`. This is the only problematic header we have.

for the headers manually added by users, after #8516 is merged, we will be able to order the rountrippers to that one does not influence the other one.

Reply:
Just to make sure I'm following properly: are you asking for `Uber-Trace-Id` to not be included in the signature or for it to be removed entirely from the request? If we're not removing it entirely, we'll need to clone the request used for generating the signature.

Reply:
My understanding is we should just not include the problematic headers in the signing, not strip them from the request.

Reply:
Yes just don't sign it.

Reply:
Thanks everyone! I've changed the round tripper to only sign a subset of headers.